File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> res;

        queue<TreeNode *> q;
        if (root) {
            q.push(root);
        }
        
        while (!q.empty()) {
            int sz = q.size();
            int levelMx = INT_MIN;
            for (int i = 0; i < sz; i++) {
                TreeNode *front = q.front();
                q.pop();
                int val = front->val;
                
                levelMx = max(levelMx, val);

                if (front->left) q.push(front->left);
                if (front->right) q.push(front->right);
            }

            res.push_back(levelMx);
        }

        return res;
    }
};

File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        if (root != null) q.offer(root);

        List<Integer> ans = new ArrayList<>();
        while (q.size() > 0) {
            int sz = q.size();

            int rowMax = Integer.MIN_VALUE;
            for (int i = 0; i < sz; i++) {
                TreeNode front = q.poll();

                rowMax = Math.max(rowMax, front.val);

                if (front.left != null) q.offer(front.left);
                if (front.right != null) q.offer(front.right);
            }

            ans.add(rowMax);
        }

        return ans;
    }
}
